// js-pack/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scriptCDN = scriptCDN;
exports.cssCDN = cssCDN;
exports.sass = sass;
exports.stylus = stylus;
exports.jsx = jsx;
exports.babel = babel;
exports.cssFile = cssFile;
exports.cssRAW = cssRAW;
exports.scriptRAW = scriptRAW;
exports.scriptFile = scriptFile;
exports.pack = pack;
const r = require('request');
const fs = require('fs');
const jquery = 'https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js';
const util = require('util');
const exec = require('child_process').exec;
const fileExists = file => fs.existsSync(file);
let node_sass = '"./node_modules/.bin/node-sass"';
let _jsx = (folder, out) => './node_modules/.bin/babel ' + folder + '  --out-dir ' + out + ' --presets=react --plugins transform-object-rest-spread,transform-es2015-classes,transform-es2015-modules-commonjs';
let _babel = (folder, out) => './node_modules/.bin/babel ' + folder + '  --out-dir ' + out + ' --plugins transform-object-rest-spread,transform-es2015-classes,transform-es2015-modules-commonjs';
let webpack = (folder, out) => './node_modules/.bin/webpack ' + folder + ' ' + out;
let _stylus = './node_modules/.bin/stylus ';
let removeFile = file => 'rm ' + file;
let removeFolder = folder => 'rm -R ' + folder;
let mkdir = folder => 'mkdir ' + folder;
let folderStart = '/';
const ran = _ => Math.floor(Math.random() * 1000 + 1);

var isWin = /^win/.test(process.platform);
if (isWin) {
  node_sass = '"./node_modules/.bin/node-sass"';
  _jsx = (folder, out) => '"./node_modules/.bin/babel" "' + folder + '"  --out-dir "' + out + '" --presets=react --plugins transform-object-rest-spread,transform-es2015-classes,transform-es2015-modules-commonjs ';
  _babel = (folder, out) => '"./node_modules/.bin/babel" "' + folder + '"  --out-dir "' + out + '" --plugins transform-object-rest-spread,transform-es2015-classes,transform-es2015-modules-commonjs';
  webpack = (folder, out) => '"./node_modules/.bin/webpack" "' + folder + '" "' + out + '"';
  _stylus = '"./node_modules/.bin/stylus" ';
  removeFile = file => 'del "' + file + '"';
  removeFolder = folder => 'rmdir /s /q "' + folder + '"';
  mkdir = folder => 'mkdir "' + folder + '"';
  folderStart = '\\';
}

const installBabel = async () => {
  if (!fs.existsSync('node_modules/babel-cli')) {
    console.log('Installing babel');

    await run('npm install --save-dev babel-cli babel-preset-react');
    await run('npm install transform-object-rest-spread');
    await run('npm install transform-es2015-classes');
    await run('npm install transform-es2015-modules-commonjs');

    console.log('babel is installed');
  }
};

const run = call => {
  return new Promise((resolve, reject) => {
    exec(call, (error, stdout, stderr) => {
      if (stderr) {
        console.log(stderr);
      }
      if (error) {
        //console.log(error)
      }
      resolve(stdout);
    });
  });
};

const downloadCDN = async (url, typeFunc) => {
  return new Promise((resolve, reject) => {
    r({ url: url }, (error, response, body) => {
      resolve(new typeFunc(body));
    });
  });
};

function Script(data) {
  this.data = data;
}

function scriptCDN(url) {
  return downloadCDN(url, Script);
}

function Style(data) {
  this.data = data;
}

function cssCDN(url) {
  return downloadCDN(url, Style);
}

async function sass(file) {
  const out = __dirname + folderStart + 'temp' + ran() + '.css';
  if (!fs.existsSync(node_sass.replace('./node', 'node').replace('"', '').replace('"', ''))) {
    console.log('install sass');
    await run('npm install node-sass --save');
    console.log('sass installed');
  }
  await run(node_sass + ' ' + file + ' ' + out);
  const data = fs.readFileSync(out, 'utf8');
  await run(removeFile(out));
  return new Style(data);
}
async function stylus(file) {
  const out = __dirname + folderStart + 'temp' + ran() + '.css';
  if (!fs.existsSync(_stylus.replace('./node', 'node').replace('"', '').replace('"', ''))) {
    console.log('Installing styles');
    await run('npm install stylus --save');
  }
  await run(_stylus + ' ' + file + ' -o ' + out);
  const data = fs.readFileSync(out, 'utf8');
  await run(removeFile(out));
  new Style(data);
}

async function jsx(folder, index) {
  const out = __dirname + folderStart + 'tempjsx' + ran();
  if (!fs.existsSync('node_modules/.bin/webpack')) {
    console.log('Installing webpack');
    await run('npm install webpack --save');
  }
  await installBabel();
  await run(mkdir(out));
  await run(_jsx(folder, out));
  await run(webpack(out + folderStart + index, out + folderStart + 'js-web-linked-packed.js'));
  const data = fs.readFileSync(out + folderStart + 'js-web-linked-packed.js', 'utf8');
  await run(removeFile(out + folderStart + 'js-web-linked-packed.js'));
  await run(removeFolder(out));
  return new Script(data);
}

async function babel(folder, index) {
  const out = __dirname + folderStart + 'tempjsx' + ran();
  if (!fs.existsSync('node_modules/.bin/webpack')) {
    console.log('Installing webpack');
    await run('npm install webpack --save');
  }
  await installBabel();
  await run(mkdir(out));
  await run(_babel(folder, out));
  await run(webpack(out + folderStart + index, out + folderStart + 'js-web-linked-packed.js'));
  const data = fs.readFileSync(out + folderStart + 'js-web-linked-packed.js', 'utf8');
  await run(removeFolder(out));
  return new Script(data);
}

function cssFile(file) {
  const data = fs.readFileSync(file, 'utf8');
  return new Style(data);
}

function cssRAW(css) {
  return new Style(css);
}

function scriptRAW(script) {
  return new Script(script);
}

function scriptFile(file) {
  const data = fs.readFileSync(file, 'utf8');
  return new Script(data);
}

const handleInjection = async injection => {
  switch (injection.constructor.name) {
    case 'AsyncFunction':
      handleInjection(injection);
      break;
    case 'Function':
      handleInjection(injection);
      break;
    case 'Promise':
      const d = await injection;
      return handleInjection(d);
      break;
    case 'Script':
      return [true, injection.data];
      break;
    case 'Style':
      return [false, injection.data];
      break;
    default:
      return [true, data];
  }
};

async function pack(arrOfInjections, scriptOut, styleOut) {
  var script = '';
  var style = '';

  for (let i = 0; i < arrOfInjections.length; i++) {
    const [isScript, data] = await handleInjection(arrOfInjections[i]);

    if (isScript) {
      script += data + '\n';
    } else {
      style += data + '\n';
    }
  }

  if (script.trim() !== '') {
    fs.writeFileSync(scriptOut, script);
  }

  if (style.trim() !== '') {
    fs.writeFileSync(styleOut, style);
  }
}
